### PR TITLE
allow end user to add watch paths

### LINKED ISSUE #6: Watch: allow end user to add watch path
```rust
#[derive(StructOpt)]
struct Watch {
    #[structopt(long, short = "w")]
    watch_paths: Vec<PathBuf>,
}
```

### PR DESCRIPTION
**Pull request by yozhgoor**
close #6

### CODE REVIEW

File: LICENSE.Apache-2.0 [hunk 1]
```diff
@@ -186,7 +186,7 @@
       same "printed page" as the copyright notice for easier
       identification within third-party archives.
 
-   Copyright [yyyy] [name of copyright owner]
+   Copyright 2021 RustMinded
```
Comment (cecton) [inline LICENSE.Apache-2.0:189]: oh I missed that one

Comment (cecton) [reply LICENSE.Apache-2.0:189]: probably the same issue exists on gptman :D wanna make a contrib?

File: src/lib.rs [hunk 2]
```diff
@@ -215,13 +220,24 @@ fn respond_to_request(stream: &mut TcpStream, build_dir_path: impl AsRef<Path>)
 }
 
 #[derive(Debug, StructOpt)]
-pub struct Watch {}
+pub struct Watch {
+    #[structopt(long, short = "w")]
```
Comment (cecton) [inline src/lib.rs:224]:
```suggestion
    #[structopt(long = "watch", short = "w")]
```

File: src/lib.rs [hunk 3]
```diff
@@ -215,13 +220,24 @@ fn respond_to_request(stream: &mut TcpStream, build_dir_path: impl AsRef<Path>)
...
 
 impl Watch {
+    pub fn new(watch_paths: Vec<impl AsRef<Path>>) -> Self {
+        Self {
+            watch_paths: watch_paths
+                .iter()
+                .map(|path| path.as_ref().to_owned())
+                .collect(),
+        }
+    }
```
Comment (cecton) [inline src/lib.rs:236]: I guess it's time we introduce builders... I thought we would at some point. Check [typed-builder](https://crates.io/crates/typed-builder). Try to write one by hand if you are not sure what a builder is.

File: src/lib.rs [hunk 4]
```diff
@@ -215,13 +220,24 @@ fn respond_to_request(stream: &mut TcpStream, build_dir_path: impl AsRef<Path>)
...
     pub fn execute(
         &self,
-        watch_path: impl AsRef<Path>,
         mut command: process::Command,
+        build_dir_path: impl AsRef<Path>,
```
Comment (cecton) [inline src/lib.rs:240]: Oh yeah it wasn't a watch path, it was the opposite: an exclusion path.

So maybe you can generalize this by adding a field `ignore_paths` in `Watch` that would match exactly like cargo-watch:

```
    -i, --ignore <pattern>...    Ignore a glob/gitignore-style pattern
```

Try first with normal path to ignore before doing the globs. We can do the globs in another pr

File: src/lib.rs [hunk 5]
```diff
@@ -232,12 +248,19 @@ impl Watch {
...
+        let build_path = &metadata.workspace_root.as_std_path().join(build_dir_path);
 
         watcher
             .watch(&metadata.workspace_root, RecursiveMode::Recursive)
             .context("cannot watch this crate")?;
 
+        for path in &self.watch_paths {
+            match watcher.watch(path, RecursiveMode::Recursive) {
+                Ok(()) => {}
+                Err(err) => log::error!("cannot watch \"{:?}\": {}", path, err),
```
Comment (cecton) [inline src/lib.rs:260]: You forgot the api of std::path? xD
```suggestion
                Err(err) => log::error!("cannot watch {}: {}", path.display(), err),
```

File: src/lib.rs [hunk 6]
```diff
@@ -215,14 +220,44 @@ fn respond_to_request(stream: &mut TcpStream, build_dir_path: impl AsRef<Path>)
...
+            .map(|x| path.starts_with(x))
+            .any(|x| x)
+    }
+
+    pub fn is_hidden_path(&mut self, path: &Path) -> bool {
+        path.file_name()
+            .and_then(|x| x.to_str())
+            .map(|x| x.starts_with('.'))
+            .unwrap_or(false)
+    }
```
Comment (cecton) [inline src/lib.rs:258]: The user don't need this. Make it private by removing `pub`.
```suggestion
    fn is_excluded_path(&mut self, path: &Path) -> bool {
        self.exclude_paths
            .iter()
            .map(|x| path.starts_with(x))
            .any(|x| x)
    }

    fn is_hidden_path(&mut self, path: &Path) -> bool {
        path.file_name()
            .and_then(|x| x.to_str())
            .map(|x| x.starts_with('.'))
            .unwrap_or(false)
    }
```

File: src/lib.rs [hunk 7]
```diff
@@ -215,14 +220,44 @@ fn respond_to_request(stream: &mut TcpStream, build_dir_path: impl AsRef<Path>)
...
+        }
+    }
+
+    pub fn add_exclude_path(&mut self, path: impl AsRef<Path>) {
+        self.exclude_paths.push(path.as_ref().to_path_buf())
+    }
+
+    pub fn add_watch_path(&mut self, path: impl AsRef<Path>) {
+        self.watch_paths.push(path.as_ref().to_path_buf())
+    }
```
Comment (cecton) [inline src/lib.rs:244]:
```suggestion
    pub fn exclude(&mut self, path: impl AsRef<Path>) {
        self.exclude_paths.push(path.as_ref().to_path_buf())
    }

    pub fn watch(&mut self, path: impl AsRef<Path>) {
        self.watch_paths.push(path.as_ref().to_path_buf())
    }
```

File: src/lib.rs [hunk 8]
```diff
@@ -131,22 +132,26 @@ impl DevServer {
...
         let handle = std::thread::spawn(move || match self.serve(build_dir_pathbuf) {
-            Ok(()) => {}
+            Ok(()) => log::trace!("starting server"),
             Err(err) => log::error!("an error occurred when starting the dev server: {}", err),
         });
 
-        let watch = Watch {};
-        match watch.execute(build_dir_path, command) {
-            Ok(()) => {}
+        match watch.execute(command) {
```
Comment (cecton) [inline src/lib.rs:148]: You need to add the build dir to the exclude list before executing.

File: examples/demo/xtask/src/main.rs [hunk 9]
```diff
@@ -39,16 +39,15 @@ fn main() -> Result<()> {
         Command::StartServer(arg) => {
             log::trace!("Starting dev server");
             arg.serve(build_dir)?;
-            log::trace!("Shutting down dev server");
         }
         Command::Serve(arg) => {
```
Comment (cecton) [inline examples/demo/xtask/src/main.rs:43]: This should be renamed DevServer

File: examples/demo/xtask/src/main.rs [hunk 10]
```diff
@@ -39,16 +39,15 @@ fn main() -> Result<()> {
         Command::StartServer(arg) => {
```
Comment (cecton) [inline examples/demo/xtask/src/main.rs:39]: This can be removed

File: src/lib.rs [hunk 11]
```diff
@@ -215,14 +220,44 @@ fn respond_to_request(stream: &mut TcpStream, build_dir_path: impl AsRef<Path>)
...
+
+    pub fn add_watch_path(&mut self, path: impl AsRef<Path>) {
+        self.watch_paths.push(path.as_ref().to_path_buf())
+    }
+
+    pub fn is_excluded_path(&mut self, path: &Path) -> bool {
+        self.exclude_paths
+            .iter()
+            .map(|x| path.starts_with(x))
+            .any(|x| x)
```
Comment (cecton) [inline src/lib.rs:250]:
```suggestion
            .any(|x| path.starts_with(x))
```

File: src/lib.rs [hunk 12]
```diff
@@ -215,14 +220,44 @@ fn respond_to_request(stream: &mut TcpStream, build_dir_path: impl AsRef<Path>)
...
+    pub fn new() -> Self {
+        Self {
+            exclude_paths: Vec::new(),
+            watch_paths: Vec::new(),
+        }
+    }
+
+    pub fn add_exclude_path(&mut self, path: impl AsRef<Path>) {
+        self.exclude_paths.push(path.as_ref().to_path_buf())
+    }
```
Comment (cecton) [inline src/lib.rs:240]: If the path is provided by the end user, it's probably relative to the current directory.

On the other hand, if the path is provided by the xtask developer and is relative, then it is relative to the workspace directory.

Therefore you should actually have 2 exclude_paths fields:
1. one that is for structopt (for the end user)
2. one that is invisible to structopt (`#[structopt(skip)]`) for the xtask developer

The build directory for example is relative to the workspace directory.

File: src/lib.rs [hunk 13]
```diff
@@ -131,22 +132,26 @@ impl DevServer {
         Ok(())
     }
 
-    pub fn watch(self, build_dir_path: impl AsRef<Path>, command: process::Command) -> Result<()> {
+    pub fn watch(
+        self,
+        build_dir_path: impl AsRef<Path>,
+        command: process::Command,
+        watch: &mut Watch,
```
Comment (cecton) [inline src/lib.rs:139]: This should probably be flattened inside DevServer

File: src/lib.rs (hunk 3 above)
Comment (cecton) [reply src/lib.rs]: After discussion we noticed it won't be really useful here. Also it adds a dependency which we want to avoid.

File: src/lib.rs [hunk 14]
```diff
@@ -106,16 +107,166 @@ impl Build {
...
+}
+
+impl Watch {
+    pub fn new() -> Self {
+        Self {
+            exclude_paths: Vec::new(),
+            watch_paths: Vec::new(),
+            workspace_exclude_paths: Vec::new(),
+        }
+    }
```
Comment (cecton) [inline src/lib.rs:127]: I don't think we need that constructor anymore

File: src/lib.rs [hunk 15]
```diff
@@ -106,16 +107,166 @@ impl Build {
...
+
+    pub fn exclude(&mut self, path: impl AsRef<Path>) {
+        self.exclude_paths.push(path.as_ref().to_path_buf())
+    }
+
+    pub fn excludes(&mut self, paths: Vec<impl AsRef<Path>>) {
```
Comment (cecton) [inline src/lib.rs:133]: The Vec here should be an impl Iterator or IntoIterator. Check the trait Extend to see how to use that in argument

File: src/lib.rs [hunk 16]
```diff
@@ -106,16 +107,166 @@ impl Build {
...
+        self.watch_paths.push(path.as_ref().to_path_buf())
+    }
+
+    pub fn watchs(&mut self, paths: Vec<impl AsRef<Path>>) {
+        for path in paths {
+            self.watch(path)
+        }
+    }
+
+    fn is_excluded_path(&mut self, path: &Path) -> bool {
```
Comment (cecton) [inline src/lib.rs:164]: All of a sudden you stopped using `impl AsRef<Path>`

File: src/lib.rs [hunk 17]
```diff
@@ -106,16 +107,166 @@ impl Build {
...
+            };
+        }
+    }
+}
+
+impl Default for Watch {
+    fn default() -> Self {
+        Self::new()
+    }
+}
```
Comment (cecton) [inline src/lib.rs:256]: No need to instantiate

File: src/lib.rs [hunk 18]
```diff
@@ -106,16 +107,166 @@ impl Build {
...
+
 #[derive(Debug, StructOpt)]
-pub struct DevServer {
+pub struct Serve {
```
Comment (cecton) [inline src/lib.rs:259]: I like the fact that this struct is called DevServer because this is what it is. But the command is usually called "serve" or "dev".

File: src/lib.rs [hunk 19]
```diff
@@ -106,16 +107,166 @@ impl Build {
...
     #[structopt(long, default_value = "127.0.0.1")]
     ip: IpAddr,
     #[structopt(long, default_value = "8000")]
     port: u16,
+    #[structopt(skip)]
+    watch: Watch,
```
Comment (cecton) [inline src/lib.rs:265]: I said flattening. Not skipping it xD 

If you flatten, the options of Watch will become available to the user which can be useful

File: src/lib.rs [hunk 20]
```diff
@@ -106,16 +107,166 @@ impl Build {
...
+
+    pub fn watch(&mut self, path: impl AsRef<Path>) {
+        self.watch_paths.push(path.as_ref().to_path_buf())
+    }
+
+    pub fn watchs(&mut self, paths: Vec<impl AsRef<Path>>) {
+        for path in paths {
+            self.watch(path)
+        }
+    }
```
Comment (cecton) [inline src/lib.rs:162]: The naming of those functions is really not great... maybe we should check around how they call that in other libraries.

Comment (cecton) [reply src/lib.rs:162]: Maybe `exclude_path`, `exclude_paths`, `watch_path`, `watch_paths` and `exclude_workspace_path`, `exclude_workspace_paths` (see the inversion here). Wdyt?

Comment (yozhgoor) [reply src/lib.rs]: True, fixed

**Review by cecton — APPROVED** [review]
